ra_cargo_watch: log errors

Veetaha opened the pull request:
Until this moment we totally ignored all the errors from cargo process. Though this is still true, but we
now try to log ones that are critical (i.e. misconfiguration errors and ignore compile errors).

This fixes #3631, and gives us a better error message to more gracefully handle the #3265
![image](https://user-images.githubusercontent.com/36276403/76958683-d7e1f080-6920-11ea-83d8-04561c11ccc4.png)

Though I think that outputting this only to `Output` channel is not enough. We should somehow warn the user that he passed wrong arguments to `cargo-watch.args`. I didn't bother looking for how to do this now, but this PR at least gives us something.

*cc* @kiljacken @matklad

matklad commented on crates/ra_cargo_watch/src/lib.rs:
```diff
@@ -316,6 +316,8 @@ impl WatchThread {
         }
         args.extend(options.args.iter().cloned());
+        args.sort();
+        args.dedup();
```
I don't think we should dedup flags. Dedupe flag is an error, it should cause a failure.

Veetaha replied:
So do we remove the deduplication at all?

matklad commented on crates/ra_cargo_watch/src/lib.rs:
```diff
@@ -316,6 +316,8 @@ impl WatchThread {
             args.push("--all-targets".to_string());
         }
         args.extend(options.args.iter().cloned());
+        args.sort();
```
This will mess up `--package foo`, we should not sort flags. In general, a good rule of thumb is to use input things as is. For example, `cargo` deliberately avoids canonicalizing and/or normalizing paths received from user.

Veetaha commented:
Removed the dedup, left only error logging.

matklad commented on crates/ra_cargo_watch/src/lib.rs:
```diff
@@ -352,8 +352,21 @@ impl WatchThread {
+                            let output =
+                                std::str::from_utf8(&output.stderr).unwrap_or("<bad utf8 output>");
+
+                            if !output.contains("could not compile") {
```
let's rather check that exit code is equal to zero. Cargo uses `1` to mean "cli is wrong", and `101` for other errors  (like a wrong dependency). The are errors which we want to ignore, which are not "cound not compile".

Also, it seems like we should move waiting for the child into `run_cargo` function.

Veetaha replied:
@matklad 
If I move waiting for the child to `run_cargo()`, then we will send `CheckEvent::End` only after the child is released, or won't it be a problem?

matklad commented on crates/ra_cargo_watch/src/lib.rs:
```diff
@@ -259,7 +259,7 @@ pub fn run_cargo(
         .args(args)
         .stdout(Stdio::piped())
-        .stderr(Stdio::null())
+        .stderr(Stdio::piped())
```
This I believe is technically wrong, as process pipes are of bounded capacity, and if Cargo to write a lot of output to stderr, it'll block. Hm, now that I think about it, we might actually hit this case, if some proc macro writes something to stderr (because, for example, RUST_LOG env variable is set)....

Ideally, we should use `select` here to drain both pipes, but that seems like a lot of work. What if we just check the exit status, and, if it is `1`, we log the command line? I think that should be enough to diagnose the issues.

Veetaha replied:
The 1 would not be enough. In the example screenshot, I provided the status is 101...

Veetaha replied:
The problem is that we lack the general error handling infrastructure for the cargo watcher...

matklad replied:
Hm, indeed. I guess the right heuristic here is "the set of messages is empty, but the return code is non-zero".

Veetaha replied:
Hm, right, I think this is a good criterion because the only successful outcome for us will be the metadata output. However, it's bad that we cannot log the actual `stderr` contents right away. Might I ask you to give a pointer to which particular `select` you referred just for my curiosity?

bjorn3 replied:
compiletest uses this: https://github.com/rust-lang/rust/blob/f4c675c476c18b1a11041193f2f59d695b126bc8/src/tools/compiletest/src/read2.rs. Maybe copy it somewhere yet another time. Or just make a crate to be shared between all users of a copy of this code.

Veetaha replied:
Thanks bjorn, that is a good source of knowledge. I'd like to avoid copying that to `rust-analyzer` repo. Maybe to `ra_vfs`?

bjorn3 replied:
> Or just make a crate to be shared between all users of a copy of this code.

I think this would be the best.

Veetaha replied:
Yeah, it would be great, though I don't feel strong enough on this topic... Unless anyone else does it :D !
For this use case, we can just use a simple solution of not reading the stderr here for now.

Veetaha commented:
Changed `run_cargo()` to return `Result<(), CargoError>`, added some comments on better error handling in cargo watcher thread and `run_cargo()` reading `stderr`

Veetaha commented on crates/ra_cargo_watch/src/lib.rs:
```diff
@@ -291,12 +304,27 @@ pub fn run_cargo(
+        Ok(exit_code) if !exit_code.success() && !read_at_least_one_message => {
+            // FIXME: Read the stderr to display the reason, see `read2()` reference in PR comment:
+            // https://github.com/rust-analyzer/rust-analyzer/pull/3632#discussion_r395605298
+            format!("the command produced no valid metadata:\n cargo {}", args.join(" "))
```
```suggestion
            format!("the command produced no valid metadata: cargo {}", args.join(" "))
```

matklad commented on crates/ra_project_model/src/cargo_workspace.rs:
```diff
@@ -263,19 +263,15 @@ pub fn load_out_dirs(
-            args.push(feature.clone());
-        }
+    } else {
+        args.extend(cargo_features.features.iter().cloned());
```
:+1:

matklad commented:
bors r+
